lib/matplotlib/categorical.py

```diff
@@ -17,56 +19,58 @@ def register():
     elif six.PY2:
         units.registry[basestring] = CategoricalConverter()
 
-    # units.registry[pandas.Categorical] = CategoricalConverter()
-
 
 class CategoricalConverter(units.ConversionInterface):
+
     @staticmethod
     def convert(value, unit, axis):
         if isinstance(value, six.string_types):
             return 0
-
+        
         vals = np.asarray(value, dtype='str')
-        uniq = np.unique(vals)
-
-        if 'nan' in uniq:
-            vals[vals == 'nan'] = -1
-            uniq = uniq[uniq != 'nan']
-
-        if '-inf' in uniq:
-            vals[vals == '-inf'] = uniq.shape[0] - 1
-            uniq = uniq[uniq != '-inf']
-
-        if 'inf' in uniq:
-            vals[vals == 'inf'] = uniq.shape[0] - 1
-            uniq = uniq[uniq != 'inf']
-
-        vmap = dict(zip(uniq, list(range(uniq.shape[0]))))
-
-        for u in uniq:
-            vals[vals == u] = vmap[u]
+        for label, loc  in axis.unit_data.items():
+            vals[vals == label] = loc
 
         return vals.astype('int')
 
     @staticmethod
     def axisinfo(unit, axis):
-        majloc = CategoricalFormatter()
-        majfmt = CategoricalLocator()
+        seq, locs = zip(*axis.unit_data.items())
+        majloc = CategoricalFormatter(locs)
+        majfmt = CategoricalLocator(seq)
```

Comment:
yeah just saw that (was caught because of a failing test!) and fixed it, will commit up